src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_form.ts

```diff
@@ -0,0 +1,198 @@
+import { get } from 'lodash';
+
+import { Form, FormData, FieldConfig, FieldsMap, FormConfig } from '../types';
+import { mapFormFields, flattenObject, unflattenObject, Subject } from '../lib';
+
+const DEFAULT_ERROR_DISPLAY_TIMEOUT = 500;
+
+export const useForm = <T = FormData>(
+  formConfig: FormConfig<T> | undefined = {}
+): { form: Form<T> } => {
+  const {
+    onSubmit,
+    schema,
+    defaultValue = {},
+    serializer = (data: any) => data,
+    deserializer = (data: any) => data,
+    options = { errorDisplayDelay: DEFAULT_ERROR_DISPLAY_TIMEOUT, stripEmptyFields: true },
+  } = formConfig;
+  const defaultValueDeserialized =
+    Object.keys(defaultValue).length === 0 ? defaultValue : deserializer(defaultValue);
+  const [isSubmitted, setSubmitted] = useState(false);
+  const [isSubmitting, setSubmitting] = useState(false);
+  const [isValid, setIsValid] = useState(true);
+  const fieldsRefs = useRef<FieldsMap>({});
+
+  // formData$ is an observable we can subscribe to in order to receive live
+  // update of the raw form data. As an observable it does not trigger any React
+  // render().
+  // The <FormDataProvider> component is the one in charge of reading this observable
+  // and updating its state to trigger the necessary view render.
+  const formData$ = useRef<Subject<T>>(new Subject<T>(flattenObject(defaultValue) as T));
+
+  // -- HELPERS
+  // ----------------------------------
+  const fieldsToArray = () => Object.values(fieldsRefs.current);
+
+  const stripEmptyFields = (fields: FieldsMap): FieldsMap => {
+    if (options.stripEmptyFields) {
+      return Object.entries(fields).reduce(
+        (acc, [key, field]) => {
+          if (field.value !== '') {
+            acc[key] = field;
+          }
+          return acc;
+        },
+        {} as FieldsMap
+      );
+    }
+    return fields;
+  };
+
+  // -- API
+  // ----------------------------------
+  const getFormData: Form<T>['getFormData'] = (getDataOptions = { unflatten: true }) =>
+    getDataOptions.unflatten
+      ? (unflattenObject(
+          mapFormFields(stripEmptyFields(fieldsRefs.current), field => field.__serializeOutput())
+        ) as T)
+      : Object.entries(fieldsRefs.current).reduce(
+          (acc, [key, field]) => ({
+            ...acc,
+            [key]: field.__serializeOutput(),
+          }),
+          {} as T
+        );
+
+  const updateFormDataAt: Form<T>['__updateFormDataAt'] = (path, value) => {
+    const currentFormData = formData$.current.value;
+    formData$.current.next({ ...currentFormData, [path]: value });
+    return formData$.current.value;
+  };
+
+  const validateFields: Form<T>['__validateFields'] = async fieldNames => {
+    // Fields that are dirty have already been validated when their value changed. Thus their errors array
+    // already contains any possible error.
+    // So when *no* fieldNames is provided, we only validate fields that haven't been changed by the user.
```

Comment:
Hey Seb, I'm sorry I'm still having trouble wrapping my head around this. Could you explain step by step the state changes made to a field when a user enters input, and how those state changes affect whether the field is validated/not validated here? "Only validate fields that haven't been changed by the user" is just not making sense to me... isn't the point of validation to check if a user's changes are valid or not? Maybe we should Zoom?

Reply:
When a field changes, this method is called with a fieldNames array. And it only then validates the fields provided in the array. If _no_ fieldNames are provided (probably called when a user clicks the sendForm button), we only validate the fields that are pristine (= that haven't been changed by the user).

We have 2 fields: "name" and "lastName"

1. A user changes the "name"
2. `form.validateFields(['name']` is called and thus the name field has the validation
3. The user clicks the sendForm button
4. `form.validateFields()` is called (no fieldNames are provided). ---> we only need to validate the "lastName" fields as "name" has already been validated

Is it clearer?

Reply:
Ahhh, I see. Your second example really clarified the behavior of the code for me. I think the desired outcome (optimizing to only validate fields which haven't been validated) makes sense to me. But I don't believe the code communicates this intention and I'm afraid it will be very easy to accidentally break this behavior.

I suggest adding a `isValidated` state to fields (forgive me if this already exists; I haven't looked at the code in awhile). I suggest setting this state to true once validation executes for a given field, and to false once the user enters input. I believe this will make the intended behavior explicit and protect against someone accidentally breaking it. What do you think?